src/sql/src/session/vars.rs

```diff
@@ -1044,16 +1050,16 @@ impl ConnectionCounter {
 
     /// The number of reserved connections available.
     pub fn reserved_remaining(&self) -> u64 {
-        // Panic because there should never be more connections than the total limit.
-        self.limit.checked_sub(self.current).expect("underflow")
+        // It's possible when reducing the limit that current number of open
+        // connections could be greater limit so just saturate at our bounds.
+        self.limit.saturating_sub(self.current)
     }
 
     /// The total limit for non-reserved connections.
     pub fn non_reserved_limit(&self) -> u64 {
-        // Panic because superuser_reserved should always be <= limit.
-        self.limit
-            .checked_sub(self.superuser_reserved)
-            .expect("underflow")
+        // It's possible when reducing the limit that current number of open
+        // connections could be greater limit so just saturate at our bounds.
+        self.limit.saturating_sub(self.superuser_reserved)
     }
```

Comment:
I don't think it should block this PR, but I'm very confused by `superuser_reserved`. What is it and why is it related to the max connection limit? How is `reserved_remaining` is completely independent of `superuser_reserved`?

Reply:
I am also confused by superuser_reserved, it's not totally clear to me how it works. I'm going to iterate on this PR once more and see if I can make it more obvious